ch8/exercise8-6/tsh-crawler/tsh-crawler.go

```diff
@@ -0,0 +1,138 @@
+package main
+
+import (
+	"bytes"
+	"io/ioutil"
+	"log"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	teleport "github.com/gravitational/teleport/lib/client"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+type job struct {
+	dir   string
+	depth int
+}
+
+type result struct {
+	dirs  []string
+	depth int
+}
+
+type Buffer struct {
+	b bytes.Buffer
+	m sync.Mutex
+}
+
+func (b *Buffer) Write(p string) (n int, err error) {
+	b.m.Lock()
+	defer b.m.Unlock()
+	return b.b.WriteString(p)
+}
+
+var directory = kingpin.Flag("directory", "directory to crawl").Required().String()
+var user = kingpin.Flag("user", "Username is the Teleport account username").String()
+var sshProxy = kingpin.Flag("proxy", "host:port the SSH proxy can be accessed at.").Required().String()
+
+func main() {
+	kingpin.Parse()
+
+	c := teleport.Config{}
+
+	path := teleport.FullProfilePath("")
+	c.LoadProfile(path, *sshProxy)
+	c.SSHProxyAddr = *sshProxy
+
+	if *user != "" {
+		c.Username = *user
+		c.HostLogin = *user
+	}
+	dirLister, err := teleport.NewClient(&c)
+	if err != nil {
+		log.Fatalf("can't create new client %v", err)
+	}
+	newCrawler := crawlerClient{DirLister: dirLister}
+	newCrawler.Crawler(*directory, *newCrawler.DirLister)
+}
+
+type crawlerClient struct {
+	DirLister *teleport.TeleportClient
+}
+
+func (tc crawlerClient) Crawler(dir string, client teleport.TeleportClient) {
```

Comment:
nit: Function names are typically imperative, e.g. `Crawl()` intead of `Crawler()`.  Then they're attached to objects such that you'd call `crawler.Crawl()`